server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -62,12 +68,31 @@ public class SnapshotsInProgress extends AbstractNamedDiffable<Custom> implement
     // keyed by repository name
     private final Map<String, ByRepo> entries;
 
+    /**
+     * IDs of nodes which are marked for removal, or which were previously marked for removal and still have running shard snapshots.
+     */
+    // When a node is marked for removal it pauses all its shard snapshots as promptly as possible. When each shard snapshot pauses it goes
+    // back to state WAITING to allow the shard to move to a different node where its snapshot can resume. However, if the removal marker is
+    // deleted before the node shuts down then we need to make sure to resume the snapshots of any remaining shards, which we do by moving
+    // all those WAITING shards back to state INIT. The problem is that the data node needs to be able to distinguish an INIT shard whose
+    // snapshot was successfully paused and now needs to be resumed from an INIT shard whose move to state WAITING has not yet been
+    // processed on the master: the latter kind of shard will move back to state WAITING in a subsequent update and so shouldn't be resumed.
+    // The solution is to avoid moving any shards back to state INIT while the node appears in this set and, conversely, only remove nodes
+    // from this set when none of their shards are in INIT state.
```

Comment:
I'm not sure I understand how to answer this without just repeating the comment, sorry. It doesn't work to use the shutdown metadata in the cluster state directly because the data node needs to distinguish INIT shards that haven't become WAITING yet from INIT shards that are resuming after being paused. Although `SnapshotsService` does indeed look at the shutdown set, it is also vitally important to look at it in `SnapshotShardsService`.

Reply:
Sorry that the comment came out very confusing. I was basically wonder whether the following part from the PR description should be part of this comment

> This is a little tricky because we might also give up on the node
removal and delete the corresponding shutdown metadata before all the
shards have moved away from the node. In that case, to avoid overlapping
aborts and restarts the master will still wait for all the shard
snapshots to stop running (moving from state INIT to SUCCESS,
FAILED, or WAITING) before resuming any snapshot work on the node.

Or are they not related?

Reply:
I see, yes, they're related. I tried to say this in the comment already but maybe it needs more detail. I added words in 13707284bc9bd58f9ff6ce3c1b48b571cedcc32f, does that help?